File: tests/fixtures/conformance_js/ts/service.ts
// TS mirror of observable_demo/service.py (and conformance_js/service.js). console.*
// is the emit surface (stdlib-logging analog); vi.spyOn is the caplog analog. Types
// present so the TS adapter's _observable_sites is exercised under annotations.

export function recordPurchase(item: string, qty: number): { item: string; qty: number } {
  console.info("purchase recorded: %s x%d", item, qty);
  return { item, qty };
}

export function computeTax(amount: number, rate: number): number {
  const tax: number = Math.round(amount * rate * 100) / 100;
  console.info("tax computed: %s", tax); // exercised, but no test observes it
  return tax;
}

export function escalate(level: number): number {
  if (level > 5) {
    console.warn("escalation: level %d", level); // never exercised
  }
  return level;
}

export function add(a: number, b: number): number {
  return a + b; // no emit — Observable does not apply
}

File: tests/fixtures/observable_demo/service.py
"""Observable-level SPIKE fixture (tests-only; never shipped).

Four nodes chosen so an emit-assert mutation must DISCRIMINATE, not merely fire:

- record_purchase: emits, and a test asserts the emission (caplog) -> proven
- compute_tax    : emits, but the test asserts only the return     -> gap-unasserted
- escalate       : emits behind a branch no test takes             -> gap-unexercised
- add            : emits nothing                                   -> n/a
"""

import logging

logger = logging.getLogger("observable_demo")


def record_purchase(item, qty):
    logger.info("purchase recorded: %s x%d", item, qty)
    return {"item": item, "qty": qty}


def compute_tax(amount, rate):
    tax = round(amount * rate, 2)
    logger.info("tax computed: %s", tax)  # exercised, but no test observes it
    return tax


def escalate(level):
    if level > 5:
        logger.warning("escalation: level %d", level)  # never exercised
    return level


def add(a, b):
    return a + b  # no emit — Observable does not apply
